ethereum/guardian/strategy.py

```diff
@@ -304,13 +307,39 @@ def __init__(self, genesis_state, key, clockwrong=False, bravery=0.92,
         # Am I byzantine?
         self.byzantine = self.crazy_bet or self.double_block_suicide < 2**80 or self.double_bet_suicide < 2**80
 
+    _joined = False
+    _joined_at_block = -1
+
+    def join(self):
+        vcode = mk_validation_code(self.key)
+        # Make the transaction to join as a Casper guardian
+        txdata = casper_ct.encode('join', [vcode])
+        tx = mk_transaction(
+            seq=self.seq,
+            gasprice=25 * 10**9,
+            gas=1000000,
+            to=CASPER,
+            value=1500 * 10**18,
+            data=txdata,
+            key=self.key,
+            create=True,
+        )
+        DEBUG('Joining Guarding Pool',
+              tx_hash=tx.hash.encode('hex'),
+              address=self.addr.encode('hex'))
+        self.add_transaction(tx)
+
+        # Icky grossness, do this better.
+        self._joined = True
+        self._joined_at_block = len(self.blocks)
+
     # Compute as many future block proposers as possible
     def add_proposers(self):
         h = len(self.finalized_hashes) - 1
         while h >= 0 and self.stateroots[h] in (None, '\x00' * 32):
             h -= 1
         state = State(self.stateroots[h] if h >= 0 else self.genesis_state_root, self.db)
-        maxh = h + ENTER_EXIT_DELAY - 1
+        maxh = self.max_finalized_height + ENTER_EXIT_DELAY - 1
```

Comment:
Why does this change lead to same results? I think both `state` and `max_finalized_height` could be different for different validators?